pkg/controllers/jobset_controller.go

```diff
@@ -913,3 +936,88 @@ func findReplicatedStatus(replicatedJobStatus []jobset.ReplicatedJobStatus, repl
 	}
 	return jobset.ReplicatedJobStatus{}
 }
+
+func (r *JobSetReconciler) deleteJobSet(ctx context.Context, js *jobset.JobSet) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	policy := metav1.DeletePropagationForeground
+	options := []client.DeleteOption{client.PropagationPolicy(policy), client.Preconditions{UID: &js.UID}}
+	log.V(2).Info("Cleaning up JobSet", "jobset", klog.KObj(js))
+
+	return r.Client.Delete(ctx, js, options...)
+}
+
+// checkIfTTLExpired checks whether a given JobSet's TTL has expired.
+func (r *JobSetReconciler) checkIfTTLExpired(ctx context.Context, jobSet *jobset.JobSet) (expiredAt *time.Time, err error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// We don't care about the JobSets that are going to be deleted
+	if jobSet.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	now := r.clock.Now()
+	remaining, expiredAt, err := timeLeft(log, jobSet, &now)
+	if err != nil {
+		return nil, err
+	}
+
+	// TTL has expired
+	if *remaining <= 0 {
```

Comment:
Can we add a nil check before dereferencing here? It may technically be safe now but I think it is good practice just in case upstream code changes in the future, for example what if the `timeLeft()` function changes and now we encounter a nil value here?

Reply:
agree